From 16ae33c41c54b42ab9f61d83b3fb280753bf742d Mon Sep 17 00:00:00 2001
From: Yongjun Zhang <yzhang@cloudera.com>
Date: Wed, 7 Feb 2018 12:58:09 -0800
Subject: [PATCH 2721/2748] HDFS-13115. In getNumUnderConstructionBlocks(),
 ignore the inodeIds for which the inodes have
 been deleted. Contributed by Yongjun Zhang.

Ref.: CDH-64802

(cherry picked from commit f491f717e9ee6b75ad5cfca48da9c6297e94a8f7)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java

Change-Id: I7a9af220e4f6805af2b73da54c6fb807cf3ac806
---
 .../hadoop/hdfs/server/namenode/LeaseManager.java  |   13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java
index 6825b89..19510d6 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/LeaseManager.java
@@ -143,14 +143,23 @@ synchronized long getNumUnderConstructionBlocks() {
       + "acquired before counting under construction blocks";
     long numUCBlocks = 0;
     for (Long id : getINodeIdWithLeases()) {
-      final INodeFile cons = fsnamesystem.getFSDirectory().getInode(id).asFile();
+    	INode inode = fsnamesystem.getFSDirectory().getInode(id);
+    	if (inode == null) {
+    	  // The inode could have been deleted after getINodeIdWithLeases() is
+    	  // called, check here, and ignore it if so
+    	  LOG.warn("Failed to find inode " + id +
+    	      " in getNumUnderConstructionBlocks().");
+    	  continue;
+    	}
+    	final INodeFile cons = inode.asFile();
       Preconditions.checkState(cons.isUnderConstruction());
       BlockInfo[] blocks = cons.getBlocks();
       if(blocks == null)
         continue;
       for(BlockInfo b : blocks) {
-        if(!b.isComplete())
+        if(!b.isComplete()) {
           numUCBlocks++;
+        }
       }
     }
     LOG.info("Number of blocks under construction: " + numUCBlocks);
-- 
1.7.9.5

